feat: macro

quangngd opened the pull request:
# Issue

https://github.com/huytd/goxkey/issues/43

# Solution

- Add new field into config data model/config file: list of macro
- Whenever  KEY_ENTER | KEY_TAB | KEY_SPACE | KEY_ESCAP is hit, check if what is displaying (`display_buffer`) is in macro table and do the replacement
- New toggle and button in main UI
![image](https://github.com/huytd/goxkey/assets/30631476/5310b59e-28b7-4d7d-8189-babcd74aad49)
- New view to add/delete macros
![image](https://github.com/huytd/goxkey/assets/30631476/f2e0096b-dc1a-4c1f-9a6c-97e8860b02ee)

huytd commented:
Thank bạn, sau khi review thì mình có một vài góp ý:

- UI ban đầu của cửa sổ macro editor bị vertical centered, nếu có thể thì nên render bình thường như list element UI của macOS
<img width="452" alt="image" src="https://github.com/huytd/goxkey/assets/613943/65c662d1-a00c-44b4-94fd-29cd485dca9b">

- Khi chạy ở chế độ debug, gõ vào 2 textbox để thêm macro rất lag, có thể do khi gõ nó trigger hàm `update` hay gì rồi
- Nút delete một từ nhìn hơi to quá so với UI, mình suggest ở đây nên dùng kí tự (⌫ `U+232B`) hoặc (× `U+00D7`)
- Có 1 bug là khi mình đặt macro cho kí tự `t` = `tôi` thì khi gõ chữ `t=<space>` nó bị chuyển thành `ttôi` (trong trường hợp này từ được lookup nên là `t=` chứ ko phải `t`)
<img width="319" alt="image" src="https://github.com/huytd/goxkey/assets/613943/cfcdc7a9-a06b-41e3-900a-802712752cde">

- Mình nghĩ nên limit các phím để trigger macro thành `KEY_TAB` và `KEY_SPACE` thôi

quangngd commented:
> * UI ban đầu của cửa sổ macro editor bị vertical centered, nếu có thể thì nên render bình thường như list element UI của macOS

Fixed, force hai text boxes và nút đóng luôn nằm ở cuối window

![image](https://github.com/huytd/goxkey/assets/30631476/91fc295e-d349-44fc-8f90-8a5da2c30acd)


> * Khi chạy ở chế độ debug, gõ vào 2 textbox để thêm macro rất lag, có thể do khi gõ nó trigger hàm `update` hay gì rồi

Fixed bằng cách optimize hàm update_active_app. Chỉ update khỉ đổi app. Hàm này lúc trước sẽ write file mỗi khi user type hoặc click bất kì cái gì --> might be a 1000x performance increase here :))

> * Nút delete một từ nhìn hơi to quá so với UI, mình suggest ở đây nên dùng kí tự (⌫ `U+232B`) hoặc (× `U+00D7`)

Updated. Dùng ×

![image](https://github.com/huytd/goxkey/assets/30631476/4959b824-23d9-4ddf-b2e6-8bfff58fbc38)


> * Có 1 bug là khi mình đặt macro cho kí tự `t` = `tôi` thì khi gõ chữ `t=<space>` nó bị chuyển thành `ttôi` (trong trường hợp này từ được lookup nên là `t=` chứ ko phải `t`)

Fixed bằng cách thêm keycode bị thiếu cho macos. `t` `=` `space` --> `t=`

> * Mình nghĩ nên limit các phím để trigger macro thành `KEY_TAB` và `KEY_SPACE` thôi

Updated.

> * Trạng thái của checkbox "Gõ tắt" chưa đuợc lưu lại trong file config

Updated.

huytd commented:
Cool, giờ còn conflict ở 2 file `main.rs` và `ui.rs` so với `main` branch, bạn fix cái này luôn nhé

quangngd commented:
resolved

huytd commented:
conflict tiếp rồi :v (do mình vừa merge cái PR bên kia)

huytd commented:
Nhân tiện khi fix conflict này thì bạn thêm 1 empty row hoặc tăng bottom padding cho cái label "Bảng gõ tắt" luôn nhé
<img width="136" alt="image" src="https://github.com/huytd/goxkey/assets/613943/9e3fd470-32ab-4854-8d1d-c7e4a7e3aaa0">

quangngd commented:
resolved
<img width="440" alt="image" src="https://github.com/huytd/goxkey/assets/30631476/46998470-243d-4fcd-b185-48e1cf83e75b">
